### ISSUE TITLE
Read config using a struct

### ISSUE THREAD

**Issue by spy16** [spf13/viper#607]
Is there a way to directly register a `struct` with viper and read configs appropriately ?

Something like demonstrated below:

```go
type config struct {
   Host string
   Port int
   Database database
}

type database struct {
   Host string
   Port int
}

cfg := config{
     Host: "localhost",
     Port: 8080,
}
viper.AutomaticEnv()
viper.SetConfigName("app")
viper.ReadInConfig()
viper.Unmarshal(&cfg)
```

I know `Unmarshal` exists but it doesn't do what i was expecting. My expectation for the above snippet was that, after the `Unmarshal` call, following would be true:

1. `cfg.Host` would be mapped to `HOST` env var, or `host` key in `app.yaml` or `app.json` etc.
2. `cfg.Database.Host` would be mapped to `DATABASE_HOST` env var, `database_host` key in `app.yaml` or `app.json` etc (or `host` sub-key of `database` key in these files)
3. Non-zero values given to the fields will be used as defaults if an overriding value not found in any config source (i.e., env var, yaml files, json files etc.) 

**Why?**

One of the projects i am working which is using `viper` has has large set of configuration options. It is becoming tedious to add `viper.GetString("config_name")` for every single config and then populate the config struct. If the above expectations were true, we could simply compose configs of different structs into one `config` struct and call `Unmarshal` with that struct.

**Comment by Nicolab** [comment]
I have the same, you are found a solution?

**Comment by sagikazarmark** [comment]
Looks like a duplicate of #584 

See my comment: https://github.com/spf13/viper/issues/584#issuecomment-451554896

TL;DR Unmarshal doesn't work with AutomaticEnv.

**Comment by Nicolab** [comment]
Thanks
